apps/emqx_durable_storage/src/emqx_ds_storage_skipstream_lts.erl

```diff
@@ -0,0 +1,739 @@
+                                true ->
+                                    inc_counter(?DS_SKIPSTREAM_LTS_HIT),
+                                    {ok, NextTS, Key, Msg0};
+                                false ->
+                                    %% Hash collision. Advance to the
+                                    %% next timestamp:
+                                    inc_counter(?DS_SKIPSTREAM_LTS_HASH_COLLISION),
+                                    {seek, NextTS + 1}
+                            end;
+                        _ ->
+                            %% This is index stream. Keep going:
+                            next_step(S, StaticIdx, CompressedTF, Iterators, NextTS, {seek, NextTS})
+                    end;
+                NextTS when NextTS > ExpectedTS, N > 0 ->
+                    inc_counter(?DS_SKIPSTREAM_LTS_MISS),
+                    %% Next index level is not what we expect.
+                    {seek, NextTS}
+            end
+    end.
+
+free_iterators(Its) ->
+    lists:foreach(
+        fun(#l{handle = IH}) ->
+            ok = rocksdb:iterator_close(IH)
+        end,
+        Its
+    ).
+
+%%%%%%%% Indexes %%%%%%%%%%
+
+mk_index(HashBytes, Static, Timestamp, Varying) ->
+    mk_index(HashBytes, Static, Timestamp, 1, Varying, []).
+
+mk_index(HashBytes, Static, Timestamp, N, [TopicLevel | Varying], Acc) ->
+    Op = {mk_key(Static, N, hash(HashBytes, TopicLevel), Timestamp), <<>>},
+    mk_index(HashBytes, Static, Timestamp, N + 1, Varying, [Op | Acc]);
+mk_index(_HashBytes, _Static, _Timestamp, _N, [], Acc) ->
+    Acc.
+
+%%%%%%%% Keys %%%%%%%%%%
+
+get_key_range(StaticIdx, WildcardIdx, Hash) ->
+    [
+        {iterate_lower_bound, mk_key(StaticIdx, WildcardIdx, Hash, 0)},
+        {iterate_upper_bound, mk_key(StaticIdx, WildcardIdx, Hash, ?max_ts)}
+    ].
+
+-spec match_ds_key(emqx_ds_lts:static_key(), binary()) -> ts() | false.
+match_ds_key(StaticIdx, Key) ->
+    match_key(StaticIdx, 0, <<>>, Key).
+
+-spec match_key(emqx_ds_lts:static_key(), wildcard_idx(), wildcard_hash(), binary()) ->
+    ts() | false.
+match_key(StaticIdx, 0, <<>>, Key) ->
+    TSz = size(StaticIdx),
+    case Key of
+        <<StaticIdx:TSz/binary, 0:?wcb, Timestamp:?tsb>> ->
+            Timestamp;
+        _ ->
+            false
+    end;
+match_key(StaticIdx, Idx, Hash, Key) when Idx > 0 ->
+    Tsz = size(StaticIdx),
+    Hsz = size(Hash),
+    case Key of
+        <<StaticIdx:Tsz/binary, Idx:?wcb, Hash:Hsz/binary, Timestamp:?tsb>> ->
+            Timestamp;
+        _ ->
+            false
+    end.
+
+-spec mk_key(emqx_ds_lts:static_key(), wildcard_idx(), wildcard_hash(), ts()) -> binary().
+mk_key(StaticIdx, 0, <<>>, Timestamp) ->
+    %% Data stream is identified by wildcard level = 0
+    <<StaticIdx/binary, 0:?wcb, Timestamp:?tsb>>;
+mk_key(StaticIdx, N, Hash, Timestamp) when N > 0 ->
+    %% Index stream:
+    <<StaticIdx/binary, N:?wcb, Hash/binary, Timestamp:?tsb>>.
```

Comment:
Speculation. If these wildcard indices are allocated other way around (i.e. `16#ffff` for the message stream, `16#fffe` for the first index, etc.), then a single iterator supposedly could be used to iterate through in the `next_loop/6` (obviously, at the cost of ability to visit indices in any order). It's unclear how much this would affect performance, but I guess it might be worth measuring.

Reply:
Interesting idea. Not sure about all the corner cases of using a single iterator, but at this stage it's very easy to change the order of indexes from incrementing to decrementing.

Reply:
P.S. Not sure how this will work. Wildcard index is in the prefix of the key, timestamp is the suffix. In order to jump to the next index level, the iterator will have to jump over all indexes and back.

Reply:
> the iterator will have to be moved backwards

Assumption is to start from the index with the smallest `N`, i.e. first from `16#fffc`, then `16#fffd`, etc.

> It also contradicts your other comment about specifying a tighter range for the iterators.

Yes, I know. They are independent comments.

Reply:
I see now, right. Jumping backwards is inevitable in this scheme.

> What does this achieve?

The same I guess: having a single iterator. After all, this is a speculation about if this can be beneficial to performance to manage a single iterator instead of many (e.g. for small batch sizes).

Reply:
Suppose we find out that creating rocksdb iterators is very expensive. With the current architecture we have ~N_clients * N_wildcard_filters iterators. If N_clients = 1000000, and N_wildcard_filters=3, then with this optimization we can go from 3000000 iterators to 1000000.

Or we can focus our efforts on different things, like RAM cache and stream events, and reduce the number of RocksDB iterators from 3000000 to 30000, because sessions will stop hammering storage.